### PR TITLE
SIMD-0123: Block Revenue Sharing: vote program piece

### PR DESCRIPTION
**Pull request by buffalojoec**
#### Problem
[SIMD-0123](https://github.com/solana-foundation/solana-improvement-documents/blob/main/proposals/0123-block-revenue-distribution.md) introduces a mechanism for validators to share block rewards with stakers by accumulating rewards in their vote account and distributing them at epoch boundaries.

The Vote program must be updated to support configuring block revenue commission, depositing rewards, and withdrawing excess lamports.

#### Summary of Changes
 Three main changes to the Vote Program, as defined by SIMD-0123:

  1. New `DepositDelegatorRewards` instruction allows a source account to deposit lamports into a vote account's `pending_delegator_rewards` field.
  2. Update `UpdateCommissionBps` to allow configuring block revenue commission in basis-points (`block_revenue_commission_bps`). This was previously not allowed, as SIMD-0291 only permitted setting `inflation_rewards_commission_bps`.
  3. Update `Withdraw` to consider the balance locked by `pending_delegator_rewards` when withdrawing. Withdrawable balance is limited to `total_lamports - pending_delegator_rewards - rent_exempt_minimum` and vote accounts cannot be closed (fully withdrawn) while `pending_delegator_rewards` > 0.

### CODE REVIEW

File: programs/vote/src/vote_processor.rs [hunk 1]
```diff
@@ -1328,47 +1373,61 @@ mod tests {
...
+            let accounts = process_instruction(
+                features,
+                &instruction_data,
+                transaction_accounts.clone(),
+                instruction_accounts.clone(),
+                Ok(()),
+            );
+            assert_eq!(get_commission_bps(&accounts[0], &kind), commission_bps);
+
+            // Same value - should pass.
```
Comment (jstarry) [inline programs/vote/src/vote_processor.rs:1495]: This uses the pre-update account state for this second instruction test rather than the post-update account state. So this test isn't checking idempotence-y as intended. Also, it would be nice to check that one commission kind doesn't get updated by the other.

File: programs/vote/src/vote_processor.rs [hunk 2]
```diff
@@ -4059,4 +4124,351 @@ mod tests {
...
+            &instruction_data,
+            vec![
+                (vote_pubkey, invalid_vote_account),
+                (source_pubkey, source_account.clone()),
+            ],
+            instruction_accounts.clone(),
+            Err(InstructionError::UninitializedAccount),
+        );
+
+        // Vote account is initialized but V3 (not V4).
```
Comment (jstarry) [inline programs/vote/src/vote_processor.rs]: Can you also test some edge case things like:
1. zero lamport deposit
    1. source not system owned
2. source == destination account
3. deposit overflow

File: feature-set/src/lib.rs [hunk 3]
```diff
@@ -2231,6 +2236,10 @@ pub static FEATURE_NAMES: LazyLock<AHashMap<Pubkey, &'static str>> = LazyLock::n
             enable_bls12_381_syscall::id(),
             "SIMD-0388: BLS12-381 syscalls",
         ),
+        (
+            block_revenue_sharing::id(),
+            "SIMD-0123: Block Revenue Sharing",
+        ),
```
Comment (jstarry) [inline feature-set/src/lib.rs:2313]: I prefer that half implemented features aren't added to the feature list. Let's hardcode `block_revenue_sharing` to `false` and remove this from the list until the runtime portion is completed. The `pub mod block_revenue_sharing` can stay though, no harm adding now.

File: programs/vote/src/vote_processor.rs (hunk 1 above)
Comment (buffalojoec) [reply programs/vote/src/vote_processor.rs:1495]: Good catch, added both.

**Review by jstarry — APPROVED** [review]

File: feature-set/src/lib.rs [hunk 4]
```diff
@@ -1291,6 +1292,10 @@ pub mod limit_instruction_accounts {
     solana_pubkey::declare_id!("DqbnFPASg7tHmZ6qfpdrt2M6MWoSeiicWPXxPhxqFCQ");
 }
 
+pub mod block_revenue_sharing {
+    solana_pubkey::declare_id!("HqUXZzYaxpbjHRCZHn8GLDCSecyCe2A7JD3An6asGdw4");
+}
```
Comment (steviez) [inline feature-set/src/lib.rs:1297]: I know this feature is intentionally disabled, but as an FYI, it not being added to the `FEATURE_NAMES` prevents the `solana feature status` from picking it up. I could see it being easy to forget later on which is why I mention it.

Something else I have seen that I kind of like is that if we absolutely must merge part of the feature in incrementally, you could declare the key to be something like this:
https://github.com/anza-xyz/agave/blob/6024f90d741e06fc2c27abdf56e2ca0bb59b7adf/feature-set/src/lib.rs#L776

Comment (buffalojoec) [reply feature-set/src/lib.rs:1297]: Okay, whichever you guys prefer on devops/releng side. Do you want me to push a PR to test-key the three Vote program features that are stubbed out?

Comment (steviez) [reply feature-set/src/lib.rs:1297]: I saw the PR to stub out so we can progress from there

Ideally, a feature implementation and key lands in one PR. Having the implementation strung out across several PR's/commits makes it harder to reason about the feature. I realize there is some nuance as different features have different complexity/context/etc
